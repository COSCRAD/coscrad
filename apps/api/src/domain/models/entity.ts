import { PartialDTO } from 'apps/api/src/types/partial-dto';
import { EntityType } from '../types/entityType';
import { isNullOrUndefined } from '../utilities/validation/is-null-or-undefined';
import { EntityId } from './types/entity-id';
import { EntityCompositeIdentifier } from './types/entityCompositeIdentifier';

// TODO Extract to a separate location
const isValidStringWithLength = (input: unknown): input is string =>
  typeof input === 'string' && input.length > 0;

/**
 *
 * TODO move this near the corresponding type. Note that this extra layer is
 * meant to safe guard us in case we want stricter validation on an `entity id`.
 * Should we change the rule, we can update this in once place, instead of replacing
 * `isValidStringWithLength` everywhere.
 */
const isValidEntityId = (input: unknown): input is EntityId =>
  isValidStringWithLength(input);

export abstract class Entity {
  readonly id: EntityId;

  abstract readonly type: EntityType;

  readonly published: boolean;

  constructor(dto: unknown) {
    // TODO Validate DTO
    if (!this.#validateDTO(dto))
      throw new Error(`Invalid dto for an entity: ${JSON.stringify(dto)}`);

    this.id = dto.id;

    this.published = typeof dto.published === 'boolean' ? dto.published : false;
  }

  getCompositeIdentifier = (): EntityCompositeIdentifier => ({
    type: this.type,
    id: this.id,
  });

  /**
   * TODO consider moving this to a `Serializable` mixin. There may be cases
   * where we want this behaviour on a non domain-model class without the
   * inheritance baggage.
   *  */
  toDTO<TEntity extends Entity>(this: TEntity): PartialDTO<TEntity> {
    return JSON.parse(JSON.stringify(this));
  }

  /**
   * TODO develop a general pattern \ directory structure for validation \ validators.
   * One idea: Return a symbol Valid | Errors[] from a validator instead of a
   * boolean flag.
   *
   * We may break the validators out into separate files. Then we will validate
   * a DTO before passing it to the constructor in our `instance factory` methods
   * in the mapping layer for the repositories.
   *  */
  #validateDTO = (dto: unknown): dto is PartialDTO<Entity> => {
    const { id } = dto as PartialDTO<Entity>;

    // A new entity will have id auto assigned by db
    if (!isNullOrUndefined(id) && !isValidEntityId(id)) return false;

    return true;
  };
}
