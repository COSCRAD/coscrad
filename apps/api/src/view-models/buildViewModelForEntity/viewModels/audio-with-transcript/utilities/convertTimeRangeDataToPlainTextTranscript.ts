import { MediaTimeRange } from 'apps/api/src/domain/models/audio-with-transcript/entities/MediaTimeRange';

export default (timeranges: MediaTimeRange[]): string =>
    timeranges.reduce(
        (accumulatedPlainText, { inPoint, outPoint, data }) =>
            accumulatedPlainText.concat(`\n[${inPoint}] ${data} [${outPoint}]`),
        ''
    );
