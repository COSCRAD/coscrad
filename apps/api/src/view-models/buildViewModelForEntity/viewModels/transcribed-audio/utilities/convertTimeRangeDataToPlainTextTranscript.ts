import { MediaTimeRange } from 'apps/api/src/domain/models/transcribed-audio/entities/MediaTimeRange';

export default (timeranges: MediaTimeRange[]): string =>
    timeranges.reduce(
        (accumulatedPlainText, { inPoint, outPoint, data }) =>
            accumulatedPlainText.concat(`\n[${inPoint}] ${data} [${outPoint}]`),
        ''
    );
