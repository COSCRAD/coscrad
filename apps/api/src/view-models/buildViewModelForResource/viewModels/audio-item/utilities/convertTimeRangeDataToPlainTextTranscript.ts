import { TranscriptItem } from '../../../../../domain/models/audio-item/entities/MediaTimeRange';

export default (timeranges: TranscriptItem[]): string =>
    timeranges.reduce(
        (accumulatedPlainText, { inPoint, outPoint, text: data }) =>
            accumulatedPlainText.concat(`\n[${inPoint}] ${data} [${outPoint}]`),
        ''
    );
