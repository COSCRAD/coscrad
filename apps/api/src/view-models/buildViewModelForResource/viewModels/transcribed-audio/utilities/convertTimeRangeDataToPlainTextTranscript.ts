import { MediaTimeRange } from '../../../../../domain/models/transcribed-audio/entities/MediaTimeRange';

export default (timeranges: MediaTimeRange[]): string =>
    timeranges.reduce(
        (accumulatedPlainText, { inPoint, outPoint, data }) =>
            accumulatedPlainText.concat(`\n[${inPoint}] ${data} [${outPoint}]`),
        ''
    );
