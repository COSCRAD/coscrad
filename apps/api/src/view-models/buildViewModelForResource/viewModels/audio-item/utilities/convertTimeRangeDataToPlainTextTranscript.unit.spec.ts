import {
    MultiLingualText,
    MultiLingualTextItemRole,
} from '../../../../../domain/common/entities/multi-lingual-text';
import { TranscriptItem } from '../../../../../domain/models/audio-item/entities/transcript-item.entity';
import convertTimeRangeDataToPlainTextTranscript from './convertTimeRangeDataToPlainTextTranscript';

// TODO- add this to the test data and import it here!
const testTranscript: TranscriptItem[] = [
    'Once upon a time, not long ago, there lived three dogs.',
    'Each of these dogs had a collar of a different color.',
    'One of these dogs did not obey his caller.',
    "This dog's collar was of a reddish color.",
    'One day the dog whose collar was a reddish color got into the collards',
    'to the spite of his caller.',
]
    .map((plainText, index) => ({
        inPoint: index * 500,
        outPoint: index * 500 + 400,
        text: new MultiLingualText({
            items: [
                {
                    text: plainText,
                    languageId: 'clc',
                    role: MultiLingualTextItemRole.original,
                },
            ],
        }),
        speakerInitials: 'FOO',
    }))
    .map((dto) => new TranscriptItem(dto));

describe('convertTimeRangeDataToPlainTextTranscript', () => {
    describe('when the input is an empty array', () => {
        it('should return an empty string', () => {
            const result = convertTimeRangeDataToPlainTextTranscript([]);

            expect(result).toBe('');
        });
    });

    describe('when given data for a sample transcript', () => {
        it('should produce the expected plain text', () => {
            const result = convertTimeRangeDataToPlainTextTranscript(testTranscript);

            expect(result).toMatchSnapshot();
        });
    });
});
