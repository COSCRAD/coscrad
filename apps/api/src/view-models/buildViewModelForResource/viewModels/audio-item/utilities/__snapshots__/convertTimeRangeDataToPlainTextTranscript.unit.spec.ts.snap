// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertTimeRangeDataToPlainTextTranscript when given data for a sample transcript should produce the expected plain text 1`] = `
"
[0] {clc}: Once upon a time, not long ago, there lived three dogs. [400]
[500] {clc}: Each of these dogs had a collar of a different color. [900]
[1000] {clc}: One of these dogs did not obey his caller. [1400]
[1500] {clc}: This dog's collar was of a reddish color. [1900]
[2000] {clc}: One day the dog whose collar was a reddish color got into the collards [2400]
[2500] {clc}: to the spite of his caller. [2900]"
`;
